app/models/labware_creators/pcr_cycles_binned_plate/csv_file/row.rb

```diff
@@ -0,0 +1,143 @@
+              numericality: { greater_than: 0, message: ->(object, _data) { COVERAGE_NEGATIVE % object } },
+              unless: -> { empty? || !submit_for_sequencing }
+    delegate :well_column, :concentration_column, :sanger_sample_id_column, :supplier_sample_name_column,
+             :input_amount_available_column, :input_amount_desired_column, :sample_volume_column, :diluent_volume_column,
+             :pcr_cycles_column, :submit_for_sequencing_column, :sub_pool_column, :coverage_column, to: :header
+
+    def initialize(row_config, header, index, row_data)
+      @row_config = row_config
+      @header = header
+      @index = index
+      @row_data = row_data
+
+      initialize_supplied_fields
+      initialize_customer_fields
+    end
+
+    def initialize_supplied_fields
+      @well                   = (@row_data[well_column] || '').strip.upcase
+      @concentration          = @row_data[concentration_column]&.strip&.to_f
+      @sanger_sample_id       = @row_data[sanger_sample_id_column]&.strip
+      @supplier_sample_name   = (@row_data[supplier_sample_name_column])&.strip
+      @input_amount_available = @row_data[input_amount_available_column]&.strip&.to_f
+    end
+
+    def initialize_customer_fields
+      @input_amount_desired   = @row_data[input_amount_desired_column]&.strip&.to_f
+      @sample_volume          = @row_data[sample_volume_column]&.strip&.to_f
+      @diluent_volume         = @row_data[diluent_volume_column]&.strip&.to_f
+      @pcr_cycles             = @row_data[pcr_cycles_column]&.strip&.to_i
+      @submit_for_sequencing  = parse_submit_for_sequencing
+      @sub_pool               = @row_data[sub_pool_column]&.strip&.to_i
+      @coverage               = @row_data[coverage_column]&.strip&.to_i
+    end
+
+    def parse_submit_for_sequencing
+      @submit_for_sequencing_string = (@row_data[submit_for_sequencing_column])&.strip&.upcase
+      @submit_for_sequencing_string == 'Y'
+    end
+
+    def to_s
+      if @well.present?
+        "row #{index + 2} [#{@well}]"
+      else
+        "row #{index + 2}"
+      end
+    end
+
+    def input_amount_desired_within_expected_range?
+      in_range?('input_amount_desired', input_amount_desired, @row_config.input_amount_desired_min, @row_config.input_amount_desired_max)
+    end
+
+    def sample_volume_within_expected_range?
+      in_range?('sample_volume', sample_volume, @row_config.sample_volume_min, @row_config.sample_volume_max)
+    end
+
+    def diluent_volume_within_expected_range?
+      in_range?('diluent_volume', diluent_volume, @row_config.diluent_volume_min, @row_config.diluent_volume_max)
+    end
+
+    def pcr_cycles_within_expected_range?
+      in_range?('pcr_cycles', pcr_cycles, @row_config.pcr_cycles_min, @row_config.pcr_cycles_max)
+    end
+
+    def submit_for_sequencing_has_expected_value?
+      return true if empty?
+
+      return true if %w[Y N].include? @submit_for_sequencing_string
```

Comment:
Just realised - shouldn't this be pulling those Y / N values from the config in 'PcrCyclesCsvFileUploadConfig' ?

Reply:
Hmm, maybe, how then would we convert those config values (any string) to boolean?